### ISSUE TITLE
swiftformat incorrectly loads data from configuration files

### ISSUE THREAD

**Issue by drspoton** [nicklockwood/SwiftFormat#1234]
I noticed that swiftformat loads a configuration from the file provided in the `--config` parameter and differently from the default `.swiftformat` configuration file.

```
swiftformat --config .swiftformat .
```

```
swiftformat .
```
The commands above will generate various changes in the repository.

The same thing refers to active rules (`--rules`)

**Comment by nicklockwood** [comment]
@drspoton I'm having trouble replicating this. Can you share your .swiftformat file? And what sort of differences are you seeing between the outputs?

**Comment by drspoton** [comment]
It looks that for `--rules` the default configuration file is just omitted.


```
$ swiftformat --version
0.49.13
```


```
$ swiftformat --rules  

 acronyms (disabled)
 andOperator
 anyObjectProtocol
 assertionFailures
 blankLinesAroundMark
 blankLinesAtEndOfScope
 blankLinesAtStartOfScope
 blankLinesBetweenImports (disabled)
 blankLinesBetweenScopes
 blockComments (disabled)
 braces
 consecutiveBlankLines
 consecutiveSpaces
 duplicateImports
 elseOnSameLine
 emptyBraces
 enumNamespaces
 extensionAccessControl
 fileHeader
 hoistPatternLet
 indent
 initCoderUnavailable
 isEmpty (disabled)
 leadingDelimiters
 linebreakAtEndOfFile
 linebreaks
 markTypes (disabled)
 modifierOrder
 numberFormatting
 organizeDeclarations (disabled)
 preferDouble (disabled)
 preferKeyPath
 redundantBackticks
 redundantBreak
 redundantClosure
 redundantExtensionACL
 redundantFileprivate
 redundantGet
 redundantInit
 redundantLet
 redundantLetError
 redundantNilInit
 redundantObjc
 redundantParens
 redundantPattern
 redundantRawValues
 redundantReturn
 redundantSelf
 redundantType
 redundantVoidReturnType
 semicolons
 sortDeclarations
 sortedImports
 sortedSwitchCases (disabled)
 spaceAroundBraces
 spaceAroundBrackets
 spaceAroundComments
 spaceAroundGenerics
 spaceAroundOperators
 spaceAroundParens
 spaceInsideBraces
 spaceInsideBrackets
 spaceInsideComments
 spaceInsideGenerics
 spaceInsideParens
 specifiers (deprecated)
 strongOutlets
 strongifiedSelf
 todos
 trailingClosures
 trailingCommas
 trailingSpace
 typeSugar
 unusedArguments
 void
 wrap
 wrapArguments
 wrapAttributes
 wrapConditionalBodies (disabled)
 wrapEnumCases (disabled)
 wrapMultilineStatementBraces
 wrapSwitchCases (disabled)
 yodaConditions
```


```
$ swiftformat --config .swiftformat --rules 

 acronyms
 andOperator
 anyObjectProtocol
 assertionFailures
 blankLinesAroundMark
 blankLinesAtEndOfScope (disabled)
 blankLinesAtStartOfScope (disabled)
 blankLinesBetweenImports
 blankLinesBetweenScopes
 blockComments (disabled)
 braces
 consecutiveBlankLines (disabled)
 consecutiveSpaces
 duplicateImports
 elseOnSameLine
 emptyBraces
 enumNamespaces
 extensionAccessControl
 fileHeader (disabled)
 hoistPatternLet
 indent
 initCoderUnavailable (disabled)
 isEmpty
 leadingDelimiters (disabled)
 linebreakAtEndOfFile
 linebreaks (disabled)
 markTypes (disabled)
 modifierOrder
 numberFormatting (disabled)
 organizeDeclarations (disabled)
 preferDouble (disabled)
 preferKeyPath
 redundantBackticks (disabled)
 redundantBreak
 redundantClosure
 redundantExtensionACL (disabled)
 redundantFileprivate
 redundantGet
 redundantInit
 redundantLet (disabled)
 redundantLetError
 redundantNilInit
 redundantObjc (disabled)
 redundantParens
 redundantPattern
 redundantRawValues
 redundantReturn
 redundantSelf
 redundantType
 redundantVoidReturnType
 semicolons (disabled)
 sortDeclarations
 sortedImports
 sortedSwitchCases (disabled)
 spaceAroundBraces
 spaceAroundBrackets (disabled)
 spaceAroundComments (disabled)
 spaceAroundGenerics (disabled)
 spaceAroundOperators (disabled)
 spaceAroundParens (disabled)
 spaceInsideBraces
 spaceInsideBrackets (disabled)
 spaceInsideComments (disabled)
 spaceInsideGenerics (disabled)
 spaceInsideParens (disabled)
 specifiers (deprecated)
 strongOutlets
 strongifiedSelf
 todos
 trailingClosures
 trailingCommas
 trailingSpace
 typeSugar
 unusedArguments
 void (disabled)
 wrap
 wrapArguments
 wrapAttributes
 wrapConditionalBodies (disabled)
 wrapEnumCases (disabled)
 wrapMultilineStatementBraces (disabled)
 wrapSwitchCases (disabled)
 yodaConditions (disabled)
```

```
$ cat .swiftformat 
--acronyms UUID
--allman false
--assetliterals visual-width
--beforemarks
--binarygrouping 4,8
--classthreshold 0
--closingparen same-line
--closurevoid remove
--commas inline
--conflictmarkers reject
--decimalgrouping 3,6
--elseposition same-line
--emptybraces no-space
--exclude Package.swift,Sources/Resources/Generated/*.swift
--exponentcase lowercase
--exponentgrouping disabled
--extensionacl on-declarations
--extensionlength 0
--extensionmark "MARK: - %t + %c"
--fractiongrouping disabled
--fragment false
--funcattributes prev-line
--groupedextension "MARK: %c"
--guardelse auto
--header ignore
--hexgrouping 4,8
--hexliteralcase uppercase
--ifdef no-indent
--importgrouping testable-last
--indent 4
--indentcase false
--indentstrings true
--lifecycle
--lineaftermarks true
--linebreaks lf
--markcategories true
--markextensions always
--marktypes always
--maxwidth 180
--modifierorder acl,override,required
--nevertrailing
--nospaceoperators
--nowrapoperators
--octalgrouping 4,8
--operatorfunc spaced
--patternlet hoist
--ranges spaced
--redundanttype inferred
--self remove
--selfrequired expect,size,log
--semicolons inline
--shortoptionals always
--smarttabs enabled
--stripunusedargs always
--swiftversion 5.5
--tabwidth unspecified
--trailingclosures
--trimwhitespace always
--typeattributes prev-line
--typemark "MARK: - %t"
--varattributes preserve
--voidtype void
--wraparguments after-first
--wrapcollections after-first
--wrapconditions after-first
--wrapparameters after-first
--wrapreturntype if-multiline
--wrapternary default
--wraptypealiases after-first
--xcodeindentation disabled
--yodaswap always
--disable blankLinesAtEndOfScope,blankLinesAtStartOfScope,consecutiveBlankLines,fileHeader,initCoderUnavailable,leadingDelimiters,linebreaks,numberFormatting,redundantBackticks,redundantExtensionACL,redundantLet,redundantObjc,semicolons,spaceAroundBrackets,spaceAroundComments,spaceAroundGenerics,spaceAroundOperators,spaceAroundParens,spaceInsideBrackets,spaceInsideComments,spaceInsideGenerics,spaceInsideParens,void,wrapMultilineStatementBraces,yodaConditions
--enable blankLinesBetweenImports,isEmpty,acronyms
```

**Comment by drspoton** [comment]
It looks that for `--rules` the default configuration file is just omitted.

```
$ swiftformat --rules  

 acronyms (disabled)
 andOperator
 anyObjectProtocol
 assertionFailures
 blankLinesAroundMark
 blankLinesAtEndOfScope
 blankLinesAtStartOfScope
 blankLinesBetweenImports (disabled)
 blankLinesBetweenScopes
 blockComments (disabled)
 braces
 consecutiveBlankLines
 consecutiveSpaces
 duplicateImports
 elseOnSameLine
 emptyBraces
 enumNamespaces
 extensionAccessControl
 fileHeader
 hoistPatternLet
 indent
 initCoderUnavailable
 isEmpty (disabled)
 leadingDelimiters
 linebreakAtEndOfFile
 linebreaks
 markTypes (disabled)
 modifierOrder
 numberFormatting
 organizeDeclarations (disabled)
 preferDouble (disabled)
 preferKeyPath
 redundantBackticks
 redundantBreak
 redundantClosure
 redundantExtensionACL
 redundantFileprivate
 redundantGet
 redundantInit
 redundantLet
 redundantLetError
 redundantNilInit
 redundantObjc
 redundantParens
 redundantPattern
 redundantRawValues
 redundantReturn
 redundantSelf
 redundantType
 redundantVoidReturnType
 semicolons
 sortDeclarations
 sortedImports
 sortedSwitchCases (disabled)
 spaceAroundBraces
 spaceAroundBrackets
 spaceAroundComments
 spaceAroundGenerics
 spaceAroundOperators
 spaceAroundParens
 spaceInsideBraces
 spaceInsideBrackets
 spaceInsideComments
 spaceInsideGenerics
 spaceInsideParens
 specifiers (deprecated)
 strongOutlets
 strongifiedSelf
 todos
 trailingClosures
 trailingCommas
 trailingSpace
 typeSugar
 unusedArguments
 void
 wrap
 wrapArguments
 wrapAttributes
 wrapConditionalBodies (disabled)
 wrapEnumCases (disabled)
 wrapMultilineStatementBraces
 wrapSwitchCases (disabled)
 yodaConditions
```


```
$ swiftformat --config .swiftformat --rules 

 acronyms
 andOperator
 anyObjectProtocol
 assertionFailures
 blankLinesAroundMark
 blankLinesAtEndOfScope (disabled)
 blankLinesAtStartOfScope (disabled)
 blankLinesBetweenImports
 blankLinesBetweenScopes
 blockComments (disabled)
 braces
 consecutiveBlankLines (disabled)
 consecutiveSpaces
 duplicateImports
 elseOnSameLine
 emptyBraces
 enumNamespaces
 extensionAccessControl
 fileHeader (disabled)
 hoistPatternLet
 indent
 initCoderUnavailable (disabled)
 isEmpty
 leadingDelimiters (disabled)
 linebreakAtEndOfFile
 linebreaks (disabled)
 markTypes (disabled)
 modifierOrder
 numberFormatting (disabled)
 organizeDeclarations (disabled)
 preferDouble (disabled)
 preferKeyPath
 redundantBackticks (disabled)
 redundantBreak
 redundantClosure
 redundantExtensionACL (disabled)
 redundantFileprivate
 redundantGet
 redundantInit
 redundantLet (disabled)
 redundantLetError
 redundantNilInit
 redundantObjc (disabled)
 redundantParens
 redundantPattern
 redundantRawValues
 redundantReturn
 redundantSelf
 redundantType
 redundantVoidReturnType
 semicolons (disabled)
 sortDeclarations
 sortedImports
 sortedSwitchCases (disabled)
 spaceAroundBraces
 spaceAroundBrackets (disabled)
 spaceAroundComments (disabled)
 spaceAroundGenerics (disabled)
 spaceAroundOperators (disabled)
 spaceAroundParens (disabled)
 spaceInsideBraces
 spaceInsideBrackets (disabled)
 spaceInsideComments (disabled)
 spaceInsideGenerics (disabled)
 spaceInsideParens (disabled)
 specifiers (deprecated)
 strongOutlets
 strongifiedSelf
 todos
 trailingClosures
 trailingCommas
 trailingSpace
 typeSugar
 unusedArguments
 void (disabled)
 wrap
 wrapArguments
 wrapAttributes
 wrapConditionalBodies (disabled)
 wrapEnumCases (disabled)
 wrapMultilineStatementBraces (disabled)
 wrapSwitchCases (disabled)
 yodaConditions (disabled)
```

**Comment by drspoton** [comment]
My case with formatting:

```swift
struct Example {
    let someId: String
}
```

This command produces correct results
```
$ swiftformat --cache ignore --config .swiftformat .
Running SwiftFormat...
Reading config file at /Users/rzeszot/Work/test/.swiftformat
SwiftFormat completed in 0.02s.
0/1 files formatted.
```

and this one is not
```
$ swiftformat --cache ignore .                      
Running SwiftFormat...
Reading config file at /Users/rzeszot/Work/test/.swiftformat
SwiftFormat completed in 0.01s.
1/1 files formatted.
```

For both commands I get the same information about used configuration file.
Without `--config` parameter the command produces wrong output, because in the configuration file there is updated `--acronyms UUID` rule

**Comment by nicklockwood** [comment]
@drspoton thanks! I'm able to reproduce it now.

The issue itself is a bit tricky. Basically it's to do with how `--acronyms` is handled. If you specify a config file then the list of acronyms you specify is exclusive, so in your case you've specified `UUID`, so that's the only acronym that's converted.

But if you use a `.swiftformat` file, the list is added to the existing default config instead of replacing it, which is why it complains about `someId`.

I'm not actually sure what the correct solution is here. Clearly they should both behave the same, but I think there's an argument for both behaviors. If it replaces the config there's no way to incrementally add additional acronyms for subfolders, but if it adds to it there's no way to remove any acronyms from the default config.

**Comment by nicklockwood** [comment]
Thoughts @calda ?

**Comment by nicklockwood** [comment]
The only sensible options I can think of are either to make the list always exclusive (which would mean that if you wanted to override `--acronyms` in a subfolder, you'd have to specify the full set), or to remove all the default acronyms (but that would make the rule a bit less convenient).

**Comment by calda** [comment]
Passing the option over the command line behaves differently from setting it in the config file? That does seem unexpected. I would expect both to work the same way as the config file does (completely override the default)

**Comment by calda** [comment]
Do all of the options with collection-like values have this problem, or just `acronyms`?

**Comment by nicklockwood** [comment]
It's all collection-like options, but I don't think it's true that you want them all to override (perhaps it's true for some/most though?)

**Comment by drspoton** [comment]
@nicklockwood I think it may be related to loading config in general. See the output from command `swiftformat --rules` with and without `--config .swiftformat`. It produces totally different rules.

**Comment by nicklockwood** [comment]
@drspoton this should be fixed in 0.50.0. Let me know if not.
